File: go/leetcode/03070.go
package main

func countSubmatrices(grid [][]int, k int) int {
	m := len(grid)
	n := len(grid[0])
	output := 0

	for row := range m {
		for col := range n {
			if row > 0 {
				grid[row][col] += grid[row-1][col]
			}
			if col > 0 {
				grid[row][col] += grid[row][col-1]
			}

			if row > 0 && col > 0 {
				grid[row][col] -= grid[row-1][col-1]
			}

			if grid[row][col] <= k {
				output += 1
			}
		}
	}

	return output
}

File: python/LeetCode/03070.py
from typing import List

def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
    m = len(grid)
    n = len(grid[0])

    mat = [[0] * n for _ in range(m)]

    output = 0
    # Let (i, j) be the bottom right corner
    for i in range(m):
        for j in range(n):
            tmp = 0
            if i > 0:
                tmp += mat[i - 1][j]
            if j > 0:
                tmp += mat[i][j - 1]
            tmp += grid[i][j]

            if i > 0 and j > 0:
                tmp -= mat[i - 1][j - 1]
            mat[i][j] = tmp
            if tmp <= k:
                output += 1

    return output
